### PR TITLE
Fixed #33061 -- Fixed handling nonexistent keys with negative deltas in incr()/decr() in memcached backends.

### PR DESCRIPTION
**Pull request by sondrelg**
This PR makes sure a `ValueError` is raised when a user tries to increment or decrement the value of a key that *does not* exist.  

According to [the ticket (33061)](https://code.djangoproject.com/ticket/33061), this may or may not have been true before, but now should always be true. 

[Here](https://docs.djangoproject.com/en/3.2/topics/cache/#django.core.caches.cache.decr) are the relevant docs stating the intended behavior. Specifically, this sentence was brought up in the ticket discussion as relevant:

> A ValueError will be raised if you attempt to increment or decrement a nonexistent cache key.

The ticket has some more context if anything is unclear 🙂

### CODE REVIEW

File: django/core/cache/backends/memcached.py [hunk 1]
```diff
@@ -105,37 +105,37 @@ def close(self, **kwargs):
...
 
     def incr(self, key, delta=1, version=None):
-        key = self.make_key(key, version=version)
-        self.validate_key(key)
+        made_key = self.make_key(key, version=version)
+        self.validate_key(made_key)
         # memcached doesn't support a negative delta
         if delta < 0:
-            return self._cache.decr(key, -delta)
+            return self.decr(key, -delta, version=version)
```
Comment (sondrelg) [inline django/core/cache/backends/memcached.py:112]: Note that we pass `key` here, not `made_key`. Please let me know if you have better suggestions for the `made_key` variable 🙂

**Comment by sondrelg** [comment]
I had a little bit of trouble running the right tests locally:

```
❯ ./tests/runtests.py cache.tests.PyMemcacheCacheTests                     
Testing against Django installed in '/Users/sondrelg/Documents/django/django'
Found 58 test(s).
Creating test database for alias 'default'...
System check identified no issues (0 silenced).
ssssssssssssssssssssssssssssssssssssssssssssssssssssssssss
----------------------------------------------------------------------
Ran 58 tests in 0.001s

OK (skipped=58)
Destroying test database for alias 'default'...
```

For anyone reviewing this, do you know (off the top of your head) how to enable a specific cache backend in the test settings?

**Comment by felixxm** [comment]
> For anyone reviewing this, do you know (off the top of your head) how to enable a specific cache backend in the test settings?

You need to install `memcached` and `pylibmc`/`pymemcache` locally, and define a `CACHES` setting that points at your `memcached` instance, see [docs](https://docs.djangoproject.com/en/3.2/internals/contributing/writing-code/unit-tests/#running-all-the-tests).

**Comment by sondrelg** [comment]
It was defining the `CACHES` setting I was a little bit confused by, but I ended up just changing `global_settings` temporarily and that at least did the trick. Thanks for the feedback 🙂 

With `PyMemcacheCache` enabled, the first commits regression tests fail as intended in 94d7c2a04d8c4136a6f4590a2722753 👍 

```
❯ git checkout 94d7c2a04d8c4136a6f4590a2722753d8e1d2f37          
Note: switching to '94d7c2a04d8c4136a6f4590a2722753d8e1d2f37'.

...

HEAD is now at 94d7c2a04d Refs #33061 -- Add regression test
```
```
❯ ./tests/runtests.py cache.tests
Testing against Django installed in '/Users/sondrelg/Documents/django/django'
Found 546 test(s).
Creating test database for alias 'default'...
Creating test database for alias 'other'...
System check identified no issues (0 silenced).
..............................................................................................................................................................................ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss...............ss..F.................F..................s.......................................................................................................................................................................................................
======================================================================
FAIL: test_decr (cache.tests.PyMemcacheCacheTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sondrelg/Documents/django/tests/cache/tests.py", line 398, in test_decr
    cache.incr('does_not_exist', -1)
AssertionError: ValueError not raised

======================================================================
FAIL: test_incr (cache.tests.PyMemcacheCacheTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sondrelg/Documents/django/tests/cache/tests.py", line 382, in test_incr
    cache.incr('does_not_exist', -1)
AssertionError: ValueError not raised

----------------------------------------------------------------------
Ran 546 tests in 86.253s

FAILED (failures=2, skipped=119)
```

File: django/core/cache/backends/memcached.py (hunk 1 above)
Comment (cjerdonek) [reply django/core/cache/backends/memcached.py:112]: Personally, I think `new_key` would be a more commonly used choice.

File: tests/cache/tests.py [hunk 2]
```diff
@@ -378,6 +378,8 @@ def test_incr(self):
         self.assertEqual(cache.incr('answer', -10), 42)
         with self.assertRaises(ValueError):
             cache.incr('does_not_exist')
+        with self.assertRaises(ValueError):
+            cache.incr('does_not_exist', -1)
```
Comment (cjerdonek) [inline tests/cache/tests.py:382]: You might as well add an assertion like this to `DummyCacheTests`'s `test_incr()` and `test_decr()` for consistency / completeness, even though it wouldn't be a regression test for your change: https://github.com/django/django/blob/5a8e8f80bb82a867eab7e4d9d099f21d0a976d22/tests/cache/tests.py#L135-L141

**Review by cjerdonek — COMMENTED** [review]
A couple comments.

File: django/core/cache/backends/memcached.py (hunk 1 above)
Comment (cjerdonek) [reply django/core/cache/backends/memcached.py:112]: Also, you might consider adding a 1-line comment before this line, for the same reason you made the comment here, e.g.

```python
# Pass the original key rather than the new key.
```

That way, readers of the code will know it's deliberate rather than an oversight.

Comment (cjerdonek) [reply django/core/cache/backends/memcached.py:112]: On second thought, it looks like you can just move the `make_key()` / `validate_key()` lines after the `delta < 0` block and not include a comment, since the new key isn't needed in that case. (Otherwise, you will be unnecessarily validating the original key twice when `delta < 0`.)

Comment (sondrelg) [reply django/core/cache/backends/memcached.py:112]: True! I'll do that 🙂

Comment (sondrelg) [reply django/core/cache/backends/memcached.py:110]: Force pushed into 0931dd244a48e8b75c24ac4b67fdee1c00d2014d

File: django/core/cache/backends/memcached.py [hunk 3]
```diff
@@ -105,14 +105,15 @@ def close(self, **kwargs):
...
-        key = self.make_key(key, version=version)
-        self.validate_key(key)
         # memcached doesn't support a negative delta
         if delta < 0:
-            return self._cache.decr(key, -delta)
+            return self.decr(key, -delta, version=version)
+
+        key = self.make_key(key, version=version)
+        self.validate_key(key)
+
```
Comment (cjerdonek) [inline django/core/cache/backends/memcached.py:114]: Remove blank line (and below).

File: django/core/cache/backends/memcached.py [hunk 4]
```diff
@@ -105,14 +105,15 @@ def close(self, **kwargs):
...
         try:
             val = self._cache.incr(key, delta)
-
```
Comment (cjerdonek) [inline django/core/cache/backends/memcached.py:115]: Undo unrelated change (and below).

File: django/core/cache/backends/memcached.py [hunk 5]
```diff
@@ -105,14 +105,15 @@ def close(self, **kwargs):
         self._cache.disconnect_all()
 
     def incr(self, key, delta=1, version=None):
-        key = self.make_key(key, version=version)
-        self.validate_key(key)
         # memcached doesn't support a negative delta
         if delta < 0:
-            return self._cache.decr(key, -delta)
+            return self.decr(key, -delta, version=version)
+
```
Comment (cjerdonek) [inline django/core/cache/backends/memcached.py:111]: Remove blank line (and below).

**Review by cjerdonek — APPROVED** [review]
I made a couple formatting requests. @felixxm may want you to do something different, given the follow-up PR he mentioned on the ticket, but I'm okay with what you have queued up. He also may want to put the `DummyCache` tests in a separate commit, but he can do that on his own.

**Comment by felixxm** [comment]
@sondrelg Thanks :1st_place_medal:  Welcome aboard :boat: 

@cjerdonek Thanks for the review :+1:
